### PR TITLE
Add option to cache fetched pipelines

### PR DESCRIPTION
**Pull request by julienduchesne**
This PR also fixes issues needed to host the plugin on jenkinsci: https://issues.jenkins-ci.org/browse/HOSTING-837

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java [hunk 1]
```diff
@@ -147,8 +188,21 @@ public String getCredentialsId() {
             return SCM._for(job);
         }
 
-        public ListBoxModel doFillCredentialsIdItems(@QueryParameter String scriptUrl, @QueryParameter String credentialsId) {
-            return new StandardListBoxModel().includeEmptyValue().includeMatchingAs(ACL.SYSTEM, Jenkins.getInstance(), StandardUsernamePasswordCredentials.class, URIRequirementBuilder.fromUri(scriptUrl).build(), CredentialsMatchers.always()).includeCurrentValue(credentialsId);
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item item, @QueryParameter String scriptUrl,
```
Comment (julienduchesne) [inline src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java]: This is what's needed for https://issues.jenkins-ci.org/browse/HOSTING-837

File: src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java [hunk 2]
```diff
@@ -110,7 +122,9 @@ public String getCredentialsId() {
 
         HttpGet httpGet = new HttpGet(expandedScriptUrl);
         if (credentialsId != null) {
-            UsernamePasswordCredentials credentials = CredentialsMatchers.firstOrNull(CredentialsProvider.lookupCredentials(UsernamePasswordCredentials.class, Jenkins.getInstance(), ACL.SYSTEM, Collections.emptyList()), CredentialsMatchers.withId(credentialsId));
+            UsernamePasswordCredentials credentials = CredentialsMatchers.firstOrNull(CredentialsProvider
```
Comment (AbdulRahmanAlHamali) [inline src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java]: what's the difference between `Jenkins.get` and `Jenkins.getInstance`?

Comment (julienduchesne) [reply src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java]: Jenkins.get is the new way. :man_shrugging: Don't ask

File: src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java [hunk 3]
```diff
@@ -119,25 +133,52 @@ public String getCredentialsId() {
...
+                if (pipelineCache.get(expandedScriptUrl).expirationDate.isBefore(Instant.now())) {
+                    listener.getLogger().println("Cache is expired. Clearing");
+                    pipelineCache.remove(expandedScriptUrl);
+                }
+            } else {
+                listener.getLogger().println("Cache miss. Actually fetching from HTTP");
+            }
+
+            if (!pipelineCache.containsKey(expandedScriptUrl)) {
+                client.connect("get pipeline",
```
Comment (AbdulRahmanAlHamali) [inline src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java]: I don't know much about the context here, but is this code asynchronous? Is it possible for whatever is after to be executed before it?

Comment (julienduchesne) [reply src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CpsHttpFlowDefinition.java]: No, it's not asynchronous but the pipelineCache can be accessed in parallel (different jobs). That's why it's a concurrent map

File: src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CacheEntry.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.jenkinsci.plugins.workflowhttp.cps;
+
+import java.time.Instant;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class CacheEntry {
```
Comment (AbdulRahmanAlHamali) [inline src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CacheEntry.java]: not so happy with the fact that we're counting on static to store this. Is there a more reasonable way to store such information in Jenkins?

**Comment by AbdulRahmanAlHamali** [comment]
LGTM overall.

I just think we need a way to either clear the cache, or at least exclude a specific file from the cache for when we are testing

File: src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CacheEntry.java (hunk 4 above)
Comment (julienduchesne) [reply src/main/java/org/jenkinsci/plugins/workflowhttp/cps/CacheEntry.java]: What do you mean? It's a ConcurrentMap. Doesn't matter that it's static, no?

**Comment by julienduchesne** [comment]
> LGTM overall.
> 
> I just think we need a way to either clear the cache, or at least exclude a specific file from the cache for when we are testing

hmm. I'll add an excluded branches wildcard attribute and add `vPR-*` in it

**Pull request merged**